### PR TITLE
Use the new Domain.join() constructor of SymPDE

### PR DESCRIPTION
**Pull request by saidctb**
Use SymPDE version 0.17.0. See SymPDE PR #130.

### CODE REVIEW

File: psydac/api/tests/build_domain.py [hunk 1]
```diff
@@ -20,17 +20,12 @@ class TransposedPolarMapping(Mapping):
...
-
-def set_interfaces(domain, interfaces):
+def create_domain(domains, interfaces, name):
+    connectivity = []
+    domains_interiors = [D.interior for D in domains]
     for I in interfaces:
-        domain = domain.join(domain, domain.name, bnd_minus=I[0], bnd_plus=I[1], direction=I[2])
-    return domain
+        connectivity.append(((domains_interiors.index(I[0].domain),I[0].axis, I[0].ext), (domains_interiors.index(I[1].domain), I[1].axis, I[1].ext), I[2]))
+    return Domain.join(domains, connectivity, name)
```
Comment (yguclu) [inline psydac/api/tests/build_domain.py:28]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/build_domain.py [hunk 2]
```diff
@@ -180,7 +175,7 @@ def build_pretzel(domain_name='pretzel', r_min=None, r_max=None):
     mapping_14_2 = TransposedPolarMapping('M14_2',2, c1= r_min+h, c2= r_min+h, rmin = 0., rmax=1.)
     domain_14_2  = mapping_14_2(dom_log_14_2)
 
-    domain = union([
+    domains = ([
                     domain_1,
```
Comment (yguclu) [inline psydac/api/tests/build_domain.py:179]: As above.

File: psydac/api/tests/build_domain.py [hunk 3]
```diff
@@ -211,6 +206,6 @@ def build_pretzel(domain_name='pretzel', r_min=None, r_max=None):
         ]
 
 
-    domain = set_interfaces(domain, interfaces)
+    domain = create_domain(domains, interfaces, domain_name)
```
Comment (yguclu) [inline psydac/api/tests/build_domain.py:209]: As above.

File: psydac/api/tests/test_2d_multipatch_mapping_maxwell.py [hunk 4]
```diff
@@ -117,9 +117,9 @@ def test_maxwell_2d_2_patch_dirichlet_0():
     D1     = mapping_1(A)
     D2     = mapping_2(B)
 
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
+    connectivity = [((0,1,1),(1,1,-1))]
+    domains = [D1,D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_maxwell.py:122]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/test_2d_multipatch_mapping_maxwell.py [hunk 5]
```diff
@@ -194,10 +194,9 @@ def test_maxwell_2d_2_patch_dirichlet_parallel_0():
     D1     = mapping_1(A)
     D2     = mapping_2(B)
 
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
-
+    connectivity = [((0,1,1),(1,1,-1))]
+    domains = [D1,D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_maxwell.py:199]: As above.

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 6]
```diff
@@ -114,9 +114,9 @@ def test_poisson_2d_2_patches_dirichlet_0():
     D1     = mapping_1(A)
     D2     = mapping_2(B)
 
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
+    connectivity = [((0,1,1),(1,1,-1))]
+    domains = [D1,D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:119]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 7]
```diff
@@ -142,9 +142,9 @@ def test_poisson_2d_2_patches_dirichlet_1():
...
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:147]: As above.

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 8]
```diff
@@ -173,13 +173,9 @@ def test_poisson_2d_2_patches_dirichlet_2():
...
-    D1D2      = D1.join(D2, name = 'D1D2',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
-
-    domain    = D1D2.join(D3, name = 'D1D2D3',
-                bnd_minus = D2.get_boundary(axis=1, ext=1),
-                bnd_plus  = D3.get_boundary(axis=1, ext=-1))
+    connectivity = [((0,1,1),(1,1,-1)), ((1,1,1),(2,1,-1))]
+    domains = [D1, D2, D3]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:178]: As above.

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 9]
```diff
@@ -281,21 +277,9 @@ def test_poisson_2d_4_patch_dirichlet_0():
...
-    domain = domain1.join(domain2, name='domain',
-                    bnd_minus = D1.get_boundary(axis=0, ext=1),
-                    bnd_plus  = D3.get_boundary(axis=0, ext=-1))
-
-    domain = domain.join(domain, name='domain',
-                        bnd_minus = D2.get_boundary(axis=0, ext=1),
-                        bnd_plus  = D4.get_boundary(axis=0, ext=-1))
+    connectivity = [((0,1,1),(1,1,-1)), ((2,1,1),(3,1,-1)), ((0,0,1),(2,0,-1)),((1,0,1),(3,0,-1))]
+    domains = [D1, D2, D3, D4]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:282]: As above.

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 10]
```diff
@@ -332,10 +316,10 @@ def test_poisson_2d_2_patches_dirichlet_parallel_0():
 
     D1     = mapping_1(A)
     D2     = mapping_2(B)
-
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
+
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:319]: Trailing white spaces.

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py [hunk 11]
```diff
@@ -332,10 +316,10 @@ def test_poisson_2d_2_patches_dirichlet_parallel_0():
...
     D1     = mapping_1(A)
     D2     = mapping_2(B)
-
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
+                
+    connectivity = [((0,1,1),(1,1,-1))]
+    domains = [D1, D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_mapping_poisson.py:322]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/test_2d_multipatch_poisson.py [hunk 12]
```diff
@@ -80,9 +80,9 @@ def test_poisson_2d_2_patch_dirichlet_0():
     A = Square('A',bounds1=(0, 0.5), bounds2=(0, 1))
     B = Square('B',bounds1=(0.5, 1.), bounds2=(0, 1))
 
-    domain = A.join(B, name = 'domain',
-                bnd_minus = A.get_boundary(axis=0, ext=1),
-                bnd_plus  = B.get_boundary(axis=0, ext=-1))
+    connectivity = [((0,0,1),(1,0,-1))]
+    domains = [A,B]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_poisson.py:85]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/test_2d_multipatch_poisson.py [hunk 13]
```diff
@@ -102,9 +102,9 @@ def test_poisson_2d_2_patch_dirichlet_1():
...
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_poisson.py:107]: As above.

File: psydac/api/tests/test_2d_multipatch_poisson.py [hunk 14]
```diff
@@ -123,9 +123,9 @@ def test_poisson_2d_2_patch_dirichlet_2():
...
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_poisson.py:128]: As above.

File: psydac/api/tests/test_2d_multipatch_poisson.py [hunk 15]
```diff
@@ -151,9 +151,10 @@ def test_poisson_2d_2_patch_dirichlet_3():
 
     D1 = M1(A)
     D2 = M2(B)
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=0, ext=1),
-                bnd_plus  = D2.get_boundary(axis=0, ext=1))
+
+    connectivity = [((0,0,1),(1,0,1))]
+    domains = [D1,D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_poisson.py:157]: As above.

File: psydac/api/tests/test_2d_multipatch_poisson.py [hunk 16]
```diff
@@ -177,9 +178,10 @@ def test_poisson_2d_2_patch_dirichlet_4():
 
     D1 = M1(A)
     D2 = M2(B)
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=0, ext=-1),
-                bnd_plus  = D2.get_boundary(axis=0, ext=-1))
+
+    connectivity = [((0,0,-1),(1,0,-1))]
+    domains = [D1,D2]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_2d_multipatch_poisson.py:184]: As above.

File: psydac/api/tests/test_postprocessing.py [hunk 17]
```diff
@@ -55,31 +55,27 @@ def build_2_mapped_squares():
     D1     = mapping_1(A)
     D2     = mapping_2(B)
 
-    domain = D1.join(D2, name = 'domain',
-                bnd_minus = D1.get_boundary(axis=1, ext=1),
-                bnd_plus  = D2.get_boundary(axis=1, ext=-1))
-    return domain
+    domains = [D1, D2]
+    connectivity = [((0,1,1),(1,1,-1))]
+    return Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_postprocessing.py:60]: Can you please rename `domains` as `patches`?

File: psydac/api/tests/test_postprocessing.py [hunk 18]
```diff
@@ -55,31 +55,27 @@ def build_2_mapped_squares():
...
     A = Square('A',bounds1=(0.5, 1.), bounds2=(0, np.pi/2))
     B = Square('B',bounds1=(0.5, 1.), bounds2=(np.pi/2, np.pi))
 
-    domain = A.join(B, 'domain',
-                    bnd_minus=A.get_boundary(1, 1),
-                    bnd_plus=B.get_boundary(1, -1))
-    return domain
+    domains = [A, B]
+    connectivity = [((0,1,1),(1,1,-1))]
+    return Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_postprocessing.py:69]: As above.

File: psydac/api/tests/test_postprocessing.py [hunk 19]
```diff
@@ -55,31 +55,27 @@ def build_2_mapped_squares():
...
     B = Cube('B',bounds1=(0.5, 1.), bounds2=(np.pi/2, np.pi), bounds3=(0, 1))
 
-    domain = A.join(B, 'domain',
-                    bnd_minus=A.get_boundary(1, 1),
-                    bnd_plus=B.get_boundary(1, -1))
-
-    return domain
+    domains = [A, B]
+    connectivity = [((0,1,1),(1,1,-1))]
+    return Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_postprocessing.py:78]: As above.

File: psydac/api/tests/test_postprocessing.py [hunk 20]
```diff
@@ -515,9 +511,9 @@ def test_reconstruct_multipatch():
     A = Square('A',bounds1=bounds1, bounds2=bounds2_A)
     B = Square('B',bounds1=bounds1, bounds2=bounds2_B)
 
-    domain = A.join(B, name='domain',
-                    bnd_minus=A.get_boundary(axis=1, ext=1),
-                    bnd_plus=B.get_boundary(axis=1, ext=-1))
+    connectivity = [((0,1,1),(1,1,-1))]
+    domains = [A,B]
+    domain = Domain.join(domains, connectivity, 'domain')
```
Comment (yguclu) [inline psydac/api/tests/test_postprocessing.py:516]: As above.

File: psydac/cad/multipatch.py
Comment (yguclu) [inline psydac/cad/multipatch.py:1]: Please replace `domains` with `patches` in this file.

File: psydac/feec/multipatch/multipatch_domain_utilities.py
Comment (yguclu) [inline psydac/feec/multipatch/multipatch_domain_utilities.py:1]: Please replace `domains` with `patches` in this file.

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good!

I have just a couple of comments. One of them is repeated many times: I would like to rename the `domains` list as `patches`, which is much clearer for the reader.

File: psydac/api/fem.py [hunk 21]
```diff
@@ -751,8 +751,13 @@ def allocate_matrices(self, backend=None):
...
 
-                        direction = target.direction
-                        direction = 1 if direction is None else direction
+                        direction = target.ornt
+                        if domain.dim == 2:
+                            assert direction == 1
+                        elif domain.dim == 3:
+                            assert all(d==1 for d in direction)
+
+                        direction = 1
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/fem.py:760]: yes , correct

File: psydac/feec/multipatch/multipatch_domain_utilities.py
Comment (saidctb) [reply psydac/feec/multipatch/multipatch_domain_utilities.py:1]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/cad/multipatch.py
Comment (saidctb) [reply psydac/cad/multipatch.py:1]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_postprocessing.py (hunk 20 above)
Comment (saidctb) [reply psydac/api/tests/test_postprocessing.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_postprocessing.py (hunk 19 above)
Comment (saidctb) [reply psydac/api/tests/test_postprocessing.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_postprocessing.py (hunk 18 above)
Comment (saidctb) [reply psydac/api/tests/test_postprocessing.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_poisson.py (hunk 16 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_poisson.py (hunk 14 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_poisson.py (hunk 13 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_poisson.py (hunk 12 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 11 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 9 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 8 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 7 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 6 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_maxwell.py (hunk 5 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_maxwell.py:199]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_maxwell.py (hunk 4 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_maxwell.py:122]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/build_domain.py (hunk 3 above)
Comment (saidctb) [reply psydac/api/tests/build_domain.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/build_domain.py (hunk 2 above)
Comment (saidctb) [reply psydac/api/tests/build_domain.py:179]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/build_domain.py (hunk 1 above)
Comment (saidctb) [reply psydac/api/tests/build_domain.py]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/fem.py [hunk 22]
```diff
@@ -790,8 +795,13 @@ def allocate_matrices(self, backend=None):
...
 
-                    direction = target.direction
-                    direction = 1 if direction is None else direction
+                    direction = target.ornt
+                    if domain.dim == 2:
+                        assert direction == 1
+                    elif domain.dim == 3:
+                        assert all(d==1 for d in direction)
+
+                    direction = 1
```
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/fem.py:806]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/fem.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/fem.py:761]: done in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

**Comment by saidctb** [comment]
> Looks good!
> 
> I have just a couple of comments. One of them is repeated many times: I would like to rename the `domains` list as `patches`, which is much clearer for the reader.

It's fixed in [d66a68e](https://github.com/pyccel/psydac/pull/309/commits/d66a68ed7d030e41dc86823c805ed685b78d906e)

File: psydac/api/tests/test_2d_multipatch_mapping_poisson.py (hunk 10 above)
Comment (saidctb) [reply psydac/api/tests/test_2d_multipatch_mapping_poisson.py]: fixed in [d7a5053](https://github.com/pyccel/psydac/pull/309/commits/d7a5053aafcd8a4ddd8ed030763825cf93691868)

**Review by yguclu — APPROVED** [review]

**Pull request merged**